File: web/src/lib/weather.ts
import {
  Cloud,
  CloudDrizzle,
  CloudFog,
  CloudLightning,
  CloudRain,
  CloudSnow,
  CloudSun,
  Sun,
  Thermometer,
  type LucideIcon,
} from 'lucide-react';

const BASE_URL = process.env.NEXT_PUBLIC_WEATHER_API_URL ?? 'https://api.open-meteo.com/v1/forecast';
const LAT = process.env.NEXT_PUBLIC_WEATHER_LAT ?? '42.5048';
const LON = process.env.NEXT_PUBLIC_WEATHER_LON ?? '27.4626';

export type Weather = {
  temperatureC: number;
  weatherCode: number;
  Icon: LucideIcon;
  iconLabel: string;
  precipitationProbability: number;
  asOfDate: string;
};

type OpenMeteoResponse = {
  current?: {
    time?: string;
    temperature_2m?: number;
    weather_code?: number;
  };
  daily?: {
    time?: string[];
    precipitation_probability_max?: (number | null)[];
  };
};

export function weatherCodeToIcon(code: number): { Icon: LucideIcon; label: string } {
  if (code === 0) return { Icon: Sun, label: 'Clear' };
  if (code === 1) return { Icon: Sun, label: 'Mainly clear' };
  if (code >= 2 && code <= 3) return { Icon: CloudSun, label: 'Partly cloudy' };
  if (code === 45 || code === 48) return { Icon: CloudFog, label: 'Fog' };
  if (code >= 51 && code <= 57) return { Icon: CloudDrizzle, label: 'Drizzle' };
  if (code >= 61 && code <= 67) return { Icon: CloudRain, label: 'Rain' };
  if (code >= 71 && code <= 77) return { Icon: CloudSnow, label: 'Snow' };
  if (code >= 80 && code <= 82) return { Icon: CloudRain, label: 'Showers' };
  if (code >= 95 && code <= 99) return { Icon: CloudLightning, label: 'Thunderstorm' };
  if (code >= 4) return { Icon: Cloud, label: 'Cloudy' };
  return { Icon: Thermometer, label: 'Weather' };
}

export async function getWeather(opts: { signal?: AbortSignal } = {}): Promise<Weather> {
  const params = new URLSearchParams({
    latitude: LAT,
    longitude: LON,
    current: 'temperature_2m,weather_code',
    daily: 'precipitation_probability_max',
    forecast_days: '1',
    timezone: 'auto',
  });
  const url = `${BASE_URL}?${params.toString()}`;
  const res = await fetch(url, {
    signal: opts.signal,
    next: { revalidate: 1800 },
  });
  if (!res.ok) {
    throw new Error(`Weather API ${res.status}`);
  }
  const data = (await res.json()) as OpenMeteoResponse;
  const tempRaw = data.current?.temperature_2m;
  const code = data.current?.weather_code ?? 0;
  const time = data.current?.time ?? '';
  const probRaw = data.daily?.precipitation_probability_max?.[0];
  if (typeof tempRaw !== 'number') {
    throw new Error('Weather API: missing temperature');
  }
  const { Icon, label } = weatherCodeToIcon(code);
  return {
    temperatureC: Math.round(tempRaw),
    weatherCode: code,
    Icon,
    iconLabel: label,
    precipitationProbability: typeof probRaw === 'number' ? probRaw : 0,
    asOfDate: time.slice(0, 10),
  };
}

File: api/src/api/weather.py
"""Open-Meteo current-weather client for Burgas Bay.

Mirrors the existing `web/src/lib/weather.ts` shape so the daily email matches
what users see on the website. Returns plain dict; dispatcher fetches once per
tick and feeds the same value to every email.
"""

from __future__ import annotations

import logging
import os
from dataclasses import dataclass
from typing import Final

import httpx

log = logging.getLogger(__name__)

_BASE_URL: Final[str] = os.getenv(
    "WEATHER_API_URL", "https://api.open-meteo.com/v1/forecast"
)
_LAT: Final[str] = os.getenv("WEATHER_LAT", "42.5048")
_LON: Final[str] = os.getenv("WEATHER_LON", "27.4626")


@dataclass(frozen=True)
class CurrentWeather:
    temperature_c: int
    weather_code: int
    label: str
    emoji: str
    precipitation_probability: int
    as_of_date: str


def _code_to_label_emoji(code: int) -> tuple[str, str]:
    if code == 0:
        return "Clear", "☀"
    if code == 1:
        return "Mainly clear", "🌤"
    if 2 <= code <= 3:
        return "Partly cloudy", "⛅"
    if code in (45, 48):
        return "Fog", "🌫"
    if 51 <= code <= 57:
        return "Drizzle", "🌦"
    if 61 <= code <= 67:
        return "Rain", "🌧"
    if 71 <= code <= 77:
        return "Snow", "🌨"
    if 80 <= code <= 82:
        return "Showers", "🌧"
    if 95 <= code <= 99:
        return "Thunderstorm", "⛈"
    return "Cloudy", "☁"


async def fetch_current_weather() -> CurrentWeather | None:
    """Fetch Burgas current weather. Returns None on any failure.

    The email is still useful without weather, so callers should treat the
    weather block as optional.
    """
    params = {
        "latitude": _LAT,
        "longitude": _LON,
        "current": "temperature_2m,weather_code",
        "daily": "precipitation_probability_max",
        "forecast_days": "1",
        "timezone": "auto",
    }
    try:
        async with httpx.AsyncClient(timeout=8) as client:
            r = await client.get(_BASE_URL, params=params)
        r.raise_for_status()
        data = r.json()
    except (httpx.HTTPError, ValueError) as exc:
        log.warning("Weather fetch failed: %s", exc)
        return None

    current = data.get("current") or {}
    daily = data.get("daily") or {}
    temp = current.get("temperature_2m")
    if not isinstance(temp, (int, float)):
        log.warning("Weather response missing temperature_2m: %s", data)
        return None
    code = int(current.get("weather_code") or 0)
    label, emoji = _code_to_label_emoji(code)
    probs = daily.get("precipitation_probability_max") or []
    prob = int(probs[0]) if probs and isinstance(probs[0], (int, float)) else 0
    as_of = str(current.get("time") or "")[:10]
    return CurrentWeather(
        temperature_c=round(temp),
        weather_code=code,
        label=label,
        emoji=emoji,
        precipitation_probability=prob,
        as_of_date=as_of,
    )
